daemon/daemon_unix.go

```diff
@@ -940,65 +943,35 @@ func initBridgeDriver(controller *libnetwork.Controller, cfg config.BridgeConfig
 		return err
 	}
 
-	var (
-		deferIPv6Alloc bool
-		ipamV6Conf     *libnetwork.IpamConf
-	)
-
-	if cfg.EnableIPv6 && cfg.FixedCIDRv6 == "" {
-		return errdefs.InvalidParameter(errors.New("IPv6 is enabled for the default bridge, but no subnet is configured. Specify an IPv6 subnet using --fixed-cidr-v6"))
-	} else if cfg.FixedCIDRv6 != "" {
-		_, fCIDRv6, err := net.ParseCIDR(cfg.FixedCIDRv6)
+	var deferIPv6Alloc bool
+	var ipamV6Conf []*libnetwork.IpamConf
+	if cfg.EnableIPv6 {
+		ipamV6Conf, err = getDefaultBridgeIPAMConf(bridgeName, userManagedBridge, cfg, netlink.FAMILY_V6)
 		if err != nil {
 			return err
 		}
-
-		// In case user has specified the daemon flag --fixed-cidr-v6 and the passed network has
-		// at least 48 host bits, we need to guarantee the current behavior where the containers'
-		// IPv6 addresses will be constructed based on the containers' interface MAC address.
-		// We do so by telling libnetwork to defer the IPv6 address allocation for the endpoints
-		// on this network until after the driver has created the endpoint and returned the
-		// constructed address. Libnetwork will then reserve this address with the ipam driver.
-		ones, _ := fCIDRv6.Mask.Size()
-		deferIPv6Alloc = ones <= 80
-
-		ipamV6Conf = &libnetwork.IpamConf{
-			AuxAddresses:  make(map[string]string),
-			PreferredPool: fCIDRv6.String(),
-		}
-
-		// In case the --fixed-cidr-v6 is specified and the current docker0 bridge IPv6
-		// address belongs to the same network, we need to inform libnetwork about it, so
-		// that it can be reserved with IPAM and it will not be given away to somebody else
-		nw6List, err := ifaceAddrs(bridgeName, netlink.FAMILY_V6)
-		if err != nil {
-			return errors.Wrap(err, "list bridge IPv6 addresses failed")
-		}
-		for _, nw6 := range nw6List {
-			if fCIDRv6.Contains(nw6.IP) {
-				ipamV6Conf.Gateway = nw6.IP.String()
-				break
+		if cfg.FixedCIDRv6 != "" {
+			_, fCIDRv6, err := net.ParseCIDR(cfg.FixedCIDRv6)
+			if err != nil {
+				return err
 			}
+			// In case user has specified the daemon flag --fixed-cidr-v6 and the passed network has
+			// at least 48 host bits, we need to guarantee the current behavior where the containers'
+			// IPv6 addresses will be constructed based on the containers' interface MAC address.
+			// We do so by telling libnetwork to defer the IPv6 address allocation for the endpoints
+			// on this network until after the driver has created the endpoint and returned the
+			// constructed address. Libnetwork will then reserve this address with the ipam driver.
+			ones, _ := fCIDRv6.Mask.Size()
+			deferIPv6Alloc = ones <= 80
```

Comment:
I think it should also apply when there's no more `--fixed-cidr-v6` / `--bip` option but the bridge already exists and has a subnet assigned.

Also, before that change there was no `--bip6`, but it should now be taken into account.

You can probably solve both issues by using `ipamV6Conf.PreferredPool`.

Reply:
Done, and added tests.